Add function pointers for every type of rectangle

ldmnyblzs opened the pull request:
These changes make it possible for the application to handle all of the rendering and JPEG decoding, thus allowing GPU accelerated rendering and multi-core decoding. I'm using this on a Raspberry Pi 2 where the CPU is not powerful enough to handle VNC.

bk138 commented on rfb/rfbclient.h:
```diff
@@ -307,6 +310,9 @@ typedef struct _rfbClient {
 
 	GotCursorShapeProc GotCursorShape;
 	GotCopyRectProc GotCopyRect;
+  GotFillRectProc GotFillRect;
```
Please add those at the end of the struct to not break ABI compatibility.

bk138 commented on rfb/rfbclient.h:
```diff
@@ -307,6 +310,9 @@ typedef struct _rfbClient {
 	GotCursorShapeProc GotCursorShape;
 	GotCopyRectProc GotCopyRect;
+  GotFillRectProc GotFillRect;
+  GotBitmapProc GotBitmap;
```
Also, can you please add some inline documention what those hooks do?

bk138 commented on libvncclient/corre.c:
```diff
@@ -46,7 +46,10 @@ HandleCoRREBPP (rfbClient* client, int rx, int ry, int rw, int rh)
 	return FALSE;
 
-    FillRectangle(client, rx, ry, rw, rh, pix);
+    if(client->GotFillRect != NULL)
```
Just a thought: You could get rid of those ifs by connecting the original functions to the hooks in vncviewer.c:rfbGetClient()

dscho replied:
I agree. It would be much nicer to initialize the `GotFillRect` attribute to `FillRectangle` in `rfbGetClient()`.

bk138 commented:
Thanks! I added some comments.

bk138 commented on rfb/rfbclient.h:
```diff
@@ -373,6 +370,18 @@ typedef struct _rfbClient {
 	LockWriteToTLSProc LockWriteToTLS;
 	UnlockWriteToTLSProc UnlockWriteToTLS;
 
+        /* Hooks for custom rendering
```
Sry, I should've mentioned that we use doxygen --> https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html So you have to start comments with /**

ldmnyblzs commented:
I moved the hooks to the end of the struct and added some comments.
I tried to implement the hooks as similarly to the already existing GotCopyRect hook as possible. That one is invoked in an if-else statement. Should I connect all of these hooks (including GotCopyRect) to the original functions in vncviewer.c:rfbGetClient() instead?

bk138 commented:
@dscho what do you think?

dscho commented on libvncclient/ultra.c:
```diff
@@ -199,8 +202,11 @@ HandleUltraZipBPP (rfbClient* client, int rx, int ry, int rw, int rh)
         CopyRectangle(client, (unsigned char *)ptr, sx, sy, sw, sh);
-        ptr += ((sw * sh) * (BPP / 8));
+      ptr += ((sw * sh) * (BPP / 8));
     }
```
This is a separate change and needs to be in its own commit.

dscho commented on rfb/rfbclient.h:
```diff
@@ -373,6 +370,18 @@ typedef struct _rfbClient {
+        GotFillRectProc GotFillRect;
+        GotBitmapProc GotBitmap;
+        /** Hook for custom JPEG decoding and rendering */
+        GotJpegProc GotJpeg;
```
It would be better to amend the original commit, rather than adding the fields in the wrong place only to fix that problem in the very same PR. Let's not introduce problems unless necessary :grinning:

dscho commented on rfb/rfbclient.h:
```diff
@@ -370,13 +370,13 @@ typedef struct _rfbClient {
          * - GotFillRect: fill an area of the framebuffer with a solid color
          * - GotBitmap: copy the bitmap in the buffer into the framebuffer
-	 * The client application should either set all three of these or none!
+         * The client application should either set all three of these or none!
```
Again, this is a problem that was introduced by the same PR that fixes it. That is not necessary. We can all pretend that no problem was introduced, ever, and that we are geniuses who never introduce problems but only fix them.

In other words: why not squash this change, as well as the previous one, into the first commit?

ldmnyblzs replied:
Can you give me more detailed instructions on how to do that? This is the first time I try to contribute to a project, so I'm not familiar with the procedure.

bk138 commented:
@ldmnyblzs OK, so please connect the hooks  to the original functions in vncviewer.c:rfbGetClient() Thanks!

ldmnyblzs commented:
Oops! The original functions are defined in libvncclient/rfbproto.c. Their declarations should be put into one of the headers, but which one?

dscho commented:
> The original functions are defined in libvncclient/rfbproto.c. Their declarations should be put into one of the headers, but which one?

Since they are declared `static` and are not used in `rfbproto.c` (or in any of the files included from there, such as `hextile.c`) anymore, the definitions should move into `libvncserver.c`, because that is now the sole user.

bk138 commented on libvncclient/rre.c:
```diff
@@ -59,7 +59,8 @@ HandleRREBPP (rfbClient* client, int rx, int ry, int rw, int rh)
     subrect.h = rfbClientSwap16IfLE(subrect.h);
 
-    FillRectangle(client, rx+subrect.x, ry+subrect.y, subrect.w, subrect.h, pix);
+    client->GotFillRect(client, rx+subrect.x, ry+subrect.y, subrect.w,
```
Really picky minor thing: pls don't add a line break here.

bk138 commented on libvncclient/tight.c:
```diff
@@ -277,7 +277,8 @@ HandleTightBPP (rfbClient* client, int rx, int ry, int rw, int rh)
 	memcpy(client->buffer, &client->buffer[numRows * rowSize], extraBytes);
 
-      CopyRectangle(client, (uint8_t *)buffer2, rx, ry+rowsProcessed, rw, numRows);
+      client->GotBitmap(client, (uint8_t *)buffer2, rx, ry+rowsProcessed, rw,
```
Same here: pls no line break.
